core/processor.go

```diff
@@ -0,0 +1,544 @@
+package core
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+
+	"github.com/ethereum-optimism/optimism/indexer/config"
+	"github.com/ethereum-optimism/optimism/op-bindings/bindings"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type Processor struct {
+	log log.Logger
+
+	L1Client *ClientExt
+	L2Client *ClientExt
+
+	L1Contracts config.L1Contracts
+	L2Contracts config.L2Contracts
+}
+
+func NewProcessor(
+	log log.Logger,
+	l1Client *ClientExt,
+	l2Client *ClientExt,
+	cfg Config,
+) *Processor {
+	log = log.New("processor", "Processor")
+
+	l2Contracts := config.L2ContractsFromPredeploys()
+	return &Processor{log, l1Client, l2Client, cfg.L1Contracts, l2Contracts}
+}
+
+func (b *Processor) ProveWithdrawalTransaction(botDelegatedWithdrawToEvent *L2ContractEvent) error {
+	receipt, err := b.L2Client.TransactionReceipt(context.Background(), common.HexToHash(botDelegatedWithdrawToEvent.TransactionHash))
+	if err != nil {
+		return err
+	}
+
+	l2BlockNumber := receipt.BlockNumber
+
+	// Events flow:
+	//
+	// event[i-5]: WithdrawalInitiated
+	// event[i-4]: ETHBridgeInitiated
+	// event[i-3]: MessagePassed
+	// event[i-2]: SentMessage
+	// event[i-1]: SentMessageExtension1
+	// event[i]  : L2StandardBridgeBot.WithdrawTo
+	if botDelegatedWithdrawToEvent.LogIndex < 5 || len(receipt.Logs) < 5 {
+		return fmt.Errorf("invalid botDelegatedWithdrawToEvent: %v", botDelegatedWithdrawToEvent)
+	}
+
+	messagePassedLog := receipt.Logs[botDelegatedWithdrawToEvent.LogIndex-3]
+	sentMessageLog := receipt.Logs[botDelegatedWithdrawToEvent.LogIndex-2]
+	sentMessageExtension1Log := receipt.Logs[botDelegatedWithdrawToEvent.LogIndex-1]
+
+	sentMessageEvent, err := b.toL2CrossDomainMessengerSentMessageExtension1(sentMessageLog, sentMessageExtension1Log)
+	if err != nil {
+		return err
+	}
+	messagePassedEvent, err := b.toMessagePassed(messagePassedLog)
+	if err != nil {
+		return err
+	}
+
+	withdrawalTx, err := b.toLowLevelMessage(sentMessageEvent, messagePassedEvent)
+	if err != nil {
+		return fmt.Errorf("toLowLevelMessage err: %v", err)
+	}
+
+	hash, err := b.hashWithdrawal(withdrawalTx)
+	if err != nil {
+		return fmt.Errorf("hashWithdrawal err: %v", err)
+	}
+
+	messageSlot, err := b.hashMessageHash(hash)
+	if err != nil {
+		return fmt.Errorf("hashMesaageHash err: %v", err)
+	}
+
+	l2OutputIndex, outputProposal, err := b.getL2OutputAfter(l2BlockNumber)
```

Comment:
I believe we can use the latest L2 block number in L2OutputOracle. Can you please give it a try? If we use an earlier block, the performance of getProof will be worse. Additionally, nodes running in certain modes may remove the early trie data, making it impossible to obtain the proof.